STM32 die temperature sensor driver rework

mathieuchopstm opened the pull request:
This PR is a rework of the STM32 die temperature sensor driver.

The main reason for this PR is the upcoming introduction of STM32WB0 series, which has a few differences from other series (calibration data is 32-bit, formula is different, sensor is always enabled) - with this reworked driver, the main differences are factored out, so the series can be easily integrated.

While at it, the driver and bindings have been heavily documented, and modified to follow the Reference Manuals more closely.

This PR also fixes errors in certain SoCs' DTSI (STM32030/STM32070 and STM32100 lines, STM32F2 series) and a few MISRA Rule 20.9 violations (using `#if xxx` on potentially undefined macros, instead of `#if defined(xxx)`).

mathieuchopstm committed:
drivers: sensor: stm32_temp: enable/disable channel in separate functions

Moves the temperature sensor channel enable and selection to separate
functions. This improves the driver's readability and makes introduction
of new STM32 series easier.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
drivers: sensor: stm32_temp: convert samples in dedicated function

Factor out the ADC-sample-to-temperature conversion logic from the sensor
subsystem API implementation, and move it to a dedicated function.

This makes the driver more readable.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
drivers: sensor: stm32_temp: make driver data field names closer to RM

Rename the fields of the STM32 dietemp sensor driver data structure to
match the names in Reference Manual more closely, and add documentation
comments to all of them.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
drivers: sensor: stm32_temp: read calibration data in dedicated function

Move the reading of calibration data from manufacturing flash to a
dedicated function. This makes introduction of new STM32 series easier.

Additionally, use the sys_read16 primitive to read data instead of
manipulating and dereferencing raw manufacturing flash pointers.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
drivers: sensor: stm32_temp: update and comment conversion code

Update all the conversion code in the STM32 dietemp sensor driver to be
more readable and match the Reference Manuals. Additionally, comment all
the code to make it easy to understand.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
dts: bindings: stm32-temp*: reword bindings documentation

Improve the STM32 dietemp sensor bindings by rewording the descriptions
of bindings and properties.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
dts: st: stm32f100: add correct 'v25' property on die temp sensor

The typical value for V25 is different on the STM32F100 line compared
to other STM32F1 MCUs. Update the DTS property to the correct value.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
dts: bindings: stm32-temp*: align 'avgslope' to datasheet format

Change the STM32 Temperature Sensor bindings to accept the average slope
value in string form instead of integer. With this change, it is possible
to use the raw decimal value found in each MCU's datasheet instead of
needing to scale it (differently depending on series!). This also allows
regrouping the property in a single file to reduce duplication.

Also update all DTSI files affected by this change and the dietemp driver
to accept the property's new format.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
dts: st: stm32f030: add correct 'avgslope' property on die temp sensor

Add the missing 'avgslope' property to the DTSI for STM32F030/STM32F070.

This fixes improper results being returned by the driver: the correct
value for the average slope is 4.3mV/°C (4300 µV/°C), but the binding's
default value of 2.53mV/°C was used instead, since property was missing.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
dts: st: stm32f2: remove 'ntc' property from die temp sensor

Remove the "Negative Temperature Coefficient" attribute from the STM32F2
die temperature sensor node, as it does not correspond to the hardware.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

erwango commented on drivers/sensor/st/stm32_temp/stm32_temp.c:
```diff
@@ -121,42 +121,93 @@ static void read_calibration_data(const struct stm32_temp_config *cfg,
 
-	temperature = ((float)data->raw * adc_ref_internal(data->adc)) / cfg->calib_vrefanalog;
-	temperature -= (calib[0] >> cfg->calib_data_shift);
+	const float Sense_Data = ((float)vdda_mv / cfg->calib_vrefanalog) * data->raw;
```
Please stick with strict snake_case
```suggestion
	const float sense_data = ((float)vdda_mv / cfg->calib_vrefanalog) * data->raw;
```
Apply everywhere

erwango commented on dts/bindings/sensor/st,stm32-temp.yaml:
```diff
@@ -39,7 +39,7 @@ properties:
 
   v25:
     type: int
-    default: 760
+    required: true
```
Please move this in a different commit, along with potential impacts reflected in dtsi or a comment explaining that there is strangely no impact.

erwango reviewed (changes requested):
Great effort, but some commit clean up required.

mathieuchopstm commented:
CI failure is due to mistake in patch (`dietemp` for STM32F100 lines is not placed properly), will fix in V2.

mathieuchopstm committed:
dts: bindings: stm32-temp: force new series to define 'v25'

Set the 'v25' property of 'st,stm32-temp' to required, and remove its
default value, to ensure new series cannot be introduced without setting
the property to the correct value explicitly.

This change does not require any DTSI modification, because there are only
two files using this compatible (stm32f1.dtsi / stm32f2.dtsi), and both of
these already set 'v25' explicitly.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
dts: bindings: stm32c0-temp-cal: force new series to define 'avgslope'

Set the 'avgslope' property of 'st,stm32c0-temp-cal' to required, and
remove its default value, to ensure new series cannot be introduced without
setting the property to the correct value explicitly.

This change does not require any DTSI modification, because there are only
two files using this compatible (stm32f030.dtsi / stm32c0.dtsi), and both
of these already set 'avgslope' explicitly.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
dts: bindings: stm32-temp: force new series to define 'avgslope'

Set the 'avgslope' property of 'st,stm32-temp' to required, and remove its
default value, to ensure new series cannot be introduced without setting
the property to the correct value explicitly.

This change does not require any DTSI modification, because there are only
two files using this compatible (stm32f1.dtsi / stm32f2.dtsi), and both of
these already set 'avgslope' explicitly.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
dts: bindings: stm32-temp*: regroup 'ntc' in common file

Reduce duplication in STM32 dietemp bindings by regrouping the 'ntc'
property declared in both "st,stm32-temp" and "st,stm32c0-temp-cal"
to the shared "st,stm32-temp-common" binding.

"st,stm32-temp-cal" is also modified to block 'ntc' property on include as
no dual-calibration sensors to date require it (this could be changed later
when need arises).

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm commented:
CI failures unrelated to PR content (twister failure caused by Renesas board, Compliance Checks seems to just be broken...)

mathieuchopstm commented:
Commits 1~3 are minor bugfixes that I think are eligible for backport to 3.7 - should they be dropped from this PR and contributed by a separate PR for this purpose, if applicable?

* [`dts: st: stm32f2: remove 'ntc' property from die temp sensor`](https://github.com/zephyrproject-rtos/zephyr/pull/77610/commits/f0444ed8234eae7d4c2a904d088867db3aa0afb3)
* [`dts: st: stm32f030: add correct 'avgslope' property on die temp sensor`](https://github.com/zephyrproject-rtos/zephyr/pull/77610/commits/d4f4ab5b2341d65bee30b121ec5d9e0754f7971c)
* [`dts: st: stm32f100: add correct 'v25' property on die temp sensor`](https://github.com/zephyrproject-rtos/zephyr/pull/77610/commits/6c39d09da1dbad8339ddb42303df340dd642cffc)

erwango commented:
@MaureenHelm Kind reminder
